modules/nf-core/deseq2/differential/templates/deseq_de.r

```diff
@@ -0,0 +1,354 @@
+                opt\$blocking,
+                ') do not correspond to sample sheet columns.'
+            )
+        )
+    }
+}
+
+# Now specify the model. Use cell-means style so we can be explicit with the
+# contrasts
+
+model <- '~ 0 +'
+
+if (!is.null(opt\$blocking_variables)) {
+    model <- paste(model, paste(blocking.vars, collapse = '+'))
+}
+
+# Make sure all the appropriate variables are factors
+
+for (v in c(blocking.vars, opt\$contrast_variable)) {
+    sample.sheet[[v]] <- as.factor(sample.sheet[[v]])
+}
+
+# Variable of interest goes last, see https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs
+
+model <- paste(model, opt\$contrast_variable, sep = ' + ')
+
+################################################
+################################################
+## Run DESeq2 processes                       ##
+################################################
+################################################
+
+dds <- DESeqDataSetFromMatrix(
+    countData = round(count.table),
+    colData = sample.sheet,
+    design = as.formula(model)
+)
+
+dds <- DESeq(
+    dds,
+    test = opt\$test,
+    fitType = opt\$fit_type,
+    minReplicatesForReplace = opt\$min_replicates_for_replace,
+    useT = opt\$use_t,
+    sfType = opt\$sf_type
+)
+
+comp.results <-
+    results(
+        dds,
+        lfcThreshold = opt\$lfc_threshold,
+        altHypothesis = opt\$alt_hypothesis,
+        independentFiltering = opt\$independent_filtering,
+        alpha = opt\$alpha,
+        pAdjustMethod = opt\$p_adjust_method,
+        minmu = opt\$minmu,
+        contrast = c(
+            opt\$contrast_variable,
+            c(opt\$treatment_level, opt\$reference_level)
+        )
+    )
+
+################################################
+################################################
+## Generate outputs                           ##
+################################################
+################################################
+
+# Common function to round numerics in data frames - mostly useful for testing,
+# since it doesn't seem to be possible to make outputs entirely reproducible
+# across machienes, even with set.seed(), but the differences are tiny.
+
+prepare_results <- function(x){
+    if (opt\$round_results){
+        format(x, nsmall=8)
+    }else{
+        x
+    }
+}
+
+
+contrast.name <-
+    paste(opt\$treatment_level, opt\$reference_level, sep = "_vs_")
+cat("Saving results for ", contrast.name, " ...\n", sep = "")
+
+# Differential expression table
+
+write.table(
+    prepare_results(data.frame(gene_id = rownames(comp.results), comp.results)),
+    file = 'deseq2.results.tsv',
+    col.names = TRUE,
+    row.names = FALSE,
+    sep = '\t',
+    quote = FALSE
+)
+
+# Dispersion plot
+
+png(
+    'deseq2.dispersion.png',
+    width = 600,
+    height = 600
+)
+plotDispEsts(dds)
+dev.off()
+
+# R object for other processes to use
+
+save(dds, file = 'dds.rld.RData')
```

Comment:
If you are saving just one object, then use `saveRDS` rather than `save`. Means the file can be read in with any filename, not always having to be read in with the name of the object you saved.

Reply:
Yep, sounds good, I agree (think I only used `save()` because I saw it in use elsewhere in nf-core)